src/ctrace.cpp

```diff
@@ -0,0 +1,356 @@
+            std::vector<cpptrace::stacktrace_frame> resolved = cpp_detail::resolve_frames(frames);
+            return ctrace::c_convert(resolved);
+        } catch(...) { // NOSONAR
+            // Don't check rethrow condition, it's risky.
+            return { nullptr, 0 };
+        }
+    }
+
+    ctrace_object_trace ctrace_raw_trace_resolve_object_trace(const ctrace_raw_trace* trace) {
+        if(!trace || !trace->frames) return { nullptr, 0 };
+        try {
+            std::vector<cpptrace::frame_ptr> frames(trace->count, 0);
+            std::copy(trace->frames, trace->frames + trace->count, frames.begin());
+            std::vector<cpptrace::object_frame> obj = cpp_detail::get_frames_object_info(frames);
+            return ctrace::c_convert(obj);
+        } catch(...) { // NOSONAR
+            // Don't check rethrow condition, it's risky.
+            return { nullptr, 0 };
+        }
+    }
+
+    ctrace_stacktrace ctrace_object_trace_resolve(const ctrace_object_trace* trace) {
+        if(!trace || !trace->frames) return { nullptr, 0 };
+        try {
+            std::vector<cpptrace::frame_ptr> frames(trace->count, 0);
+            std::transform(trace->frames, trace->frames + trace->count, frames.begin(),
+            [] (const ctrace_object_frame& frame) -> cpptrace::frame_ptr {
+                return frame.raw_address;
+            });
+            std::vector<cpptrace::stacktrace_frame> resolved = cpp_detail::resolve_frames(frames);
+            return ctrace::c_convert(resolved);
+        } catch(...) { // NOSONAR
+            // Don't check rethrow condition, it's risky.
+            return { nullptr, 0 };
+        }
+    }
+
+    // ctrace::safe:
+    size_t ctrace_safe_generate_raw_trace(ctrace_frame_ptr* buffer, size_t size, size_t skip, size_t max_depth) {
+        return cpptrace::safe_generate_raw_trace(buffer, size, skip, max_depth);
+    }
+
+    void ctrace_get_safe_object_frame(ctrace_frame_ptr address, ctrace_safe_object_frame* out) {
+        // TODO: change this?
+        static_assert(sizeof(cpptrace::safe_object_frame) == sizeof(ctrace_safe_object_frame), "");
+        cpptrace::get_safe_object_frame(address, reinterpret_cast<cpptrace::safe_object_frame*>(out));
+    }
+
+    // ctrace::io:
+    ctrace_owning_string ctrace_stacktrace_to_string(const ctrace_stacktrace* trace, ctrace_bool use_color) {
+        if(!trace || !trace->frames) return ctrace::generate_owning_string("<empty trace>");
+        auto cpp_trace = ctrace::cpp_convert(trace);
+        std::string trace_string = cpp_trace.to_string(bool(use_color));
+        return ctrace::generate_owning_string(trace_string);
+    }
+
+    void ctrace_stacktrace_print(const ctrace_stacktrace* trace, FILE* to, ctrace_bool use_color) {
+        if(use_color) cpp_detail::enable_virtual_terminal_processing_if_needed();
+        ctrace::ffprintf(to, "Stack trace (most recent call first):\n");
```

Comment:
Since `cpp_convert` is called above, would it make sense to just use that here? (even though it feels like a wasteful conversion)
Alternatively maybe we create some templated function in cpptrace to do the printing in a more polymorphic way (could defer to another PR)